File: 50. Pow(x, n)/Pow(x, n).py
class Solution(object):
    m={}
    def myPow(self, x, n):
        """
        :type x: float
        :type n: int
        :rtype: float
        """
        self.m[0]=1.0
        self.m[1]=x
        if n <0:
            return 1.0/self.myPow(x,-n)
        if n in self.m:
            return self.m[n]
        if n>=2:
            n1=n/2
            n2=n-n1
            if (n1 not in self.m):
                self.m[n1]=self.myPow(x,n1)
                #print n1
            if(n2 not in self.m):
                self.m[n2]=self.myPow(x,n2)
                #print n2
            self.m[n]=self.m[n1]*self.m[n2]
            return  self.m[n]

File: 50. Pow(x, n)/Pow(x, n).java
public class Solution {
    public double myPow(double x, int n) {
        if(n >= 0)
            return helper(x, n);
        if(n != Integer.MIN_VALUE)
            return 1.0 / helper( x, -n);
        else
            return (1.0 / x)*1.0 / (helper(x, Integer.MAX_VALUE)) ;
        //return helper(x, n);
    }
    public double helper(double x, int n){
        if(n == 0)
            return 1;
        if(n == 1)
            return x;
        double half = myPow(x, n/2);
        if(n % 2 == 0)
            return half * half;
        else
            return half * half * x;
    }
      
}